core/templates/dev/head/pages/exploration_editor/statistics_tab/StatisticsTab.js

```diff
@@ -202,3 +190,64 @@ oppia.controller('StatisticsTab', [
     };
   }
 ]);
+
+oppia.factory('StateImprovementSuggestionService', [
+  'IMPROVE_TYPE_DEFAULT', 'IMPROVE_TYPE_INCOMPLETE',
+  function(
+    IMPROVE_TYPE_DEFAULT, IMPROVE_TYPE_INCOMPLETE) {
+  return {
+    // Returns an array of suggested improvements to states. Each suggestion is
+    // an object with the keys: rank, improveType, and stateName.
+    getStateImprovements: function(explorationStates, allStateStats) {
+      var rankComparator = function(lhs, rhs) {
+        return rhs.rank - lhs.rank;
+      };
+
+      var rankedStates = [];
+      var stateNames = Object.keys(explorationStates);
+      for (var i = 0; i < stateNames.length; i++) {
+        var stateName = stateNames[i];
+        var stateStats = allStateStats[stateName];
+
+        var totalEntryCount = stateStats['total_entry_count'];
+        var noAnswerSubmittedCount = stateStats['no_submitted_answer_count'];
+        var defaultAnswerCount = stateStats['num_default_answers'];
+
+        if (totalEntryCount == 0) {
+          continue;
+        }
+
+        var threshold = 0.2 * totalEntryCount;
+        var eligibleFlags = [];
+        var state = explorationStates[stateName];
+        var stateInteraction = state.interaction;
+        if (defaultAnswerCount > threshold &&
+            stateInteraction['default_outcome'] &&
+            stateInteraction['default_outcome'].dest == stateName) {
+          eligibleFlags.push({
+            'rank': defaultAnswerCount,
```

Comment:
Fair point; this code was likely introduced before we started using a linter. :) Moreover, I have **many** more linter errors to deal with after #1205 is merged into the feature branch.